File: backend/lib/html_parser.py
"""
Utility functions for parsing HTML files and extracting JSON from <script> tags.
"""

import re
import json
from pathlib import Path
from typing import Optional, Dict, Any, List
from bs4 import BeautifulSoup


def extract_json_from_script_tag(html_content: str, variable_name: Optional[str] = None) -> Optional[Dict[str, Any]]:
    """
    Extract JSON variable from a <script> tag in HTML content.
    
    Args:
        html_content: The HTML content as a string
        variable_name: Optional name of the variable to extract (e.g., "window.data" or "data")
                      If None, will try to extract any JSON object from script tags
    
    Returns:
        The parsed JSON object, or None if not found
    
    Examples:
        # Extract a specific variable: var data = {...};
        data = extract_json_from_script_tag(html, "data")
        
        # Extract window.myData = {...};
        data = extract_json_from_script_tag(html, "window.myData")
        
        # Extract any JSON object from script tags
        data = extract_json_from_script_tag(html)
    """
    if variable_name:
        # Method 1: Extract specific variable using regex
        # Pattern matches: var variable_name = {...}; or window.variable_name = {...};
        patterns = [
            # var variable_name = {...};
            rf'\bvar\s+{re.escape(variable_name)}\s*=\s*(\{{.*?\}})\s*;',
            # let variable_name = {...};
            rf'\blet\s+{re.escape(variable_name)}\s*=\s*(\{{.*?\}})\s*;',
            # const variable_name = {...};
            rf'\bconst\s+{re.escape(variable_name)}\s*=\s*(\{{.*?\}})\s*;',
            # window.variable_name = {...};
            rf'window\.{re.escape(variable_name)}\s*=\s*(\{{.*?\}})\s*;',
            # variable_name = {...};
            rf'\b{re.escape(variable_name)}\s*=\s*(\{{.*?\}})\s*;',
        ]
        
        for pattern in patterns:
            match = re.search(pattern, html_content, re.DOTALL)
            if match:
                try:
                    return json.loads(match.group(1))
                except json.JSONDecodeError:
                    continue
        
        # Method 2: Using BeautifulSoup to find script tags and search within them
        soup = BeautifulSoup(html_content, 'html.parser')
        for script in soup.find_all('script'):
            if script.string:
                for pattern in patterns:
                    match = re.search(pattern, script.string, re.DOTALL)
                    if match:
                        try:
                            return json.loads(match.group(1))
                        except json.JSONDecodeError:
                            continue
    else:
        # Extract any JSON object from script tags
        soup = BeautifulSoup(html_content, 'html.parser')
        for script in soup.find_all('script'):
            if script.string:
                # Try to find JSON objects in the script content
                # Look for patterns like {...} or [{...}]
                json_patterns = [
                    r'(\{[^{}]*(?:\{[^{}]*\}[^{}]*)*\})',  # Nested objects
                    r'(\[[^\]]*\])',  # Arrays
                ]
                
                for pattern in json_patterns:
                    matches = re.finditer(pattern, script.string, re.DOTALL)
                    for match in matches:
                        try:
                            return json.loads(match.group(1))
                        except json.JSONDecodeError:
                            continue
    
    return None


def extract_all_json_from_script_tags(html_content: str) -> List[Dict[str, Any]]:
    """
    Extract all JSON objects from <script> tags in HTML content.
    
    Args:
        html_content: The HTML content as a string
    
    Returns:
        List of all parsed JSON objects found in script tags
    """
    results = []
    soup = BeautifulSoup(html_content, 'html.parser')
    
    for script in soup.find_all('script'):
        if script.string:
            # Find all JSON-like objects
            json_pattern = r'(\{[^{}]*(?:\{[^{}]*\}[^{}]*)*\})'
            matches = re.finditer(json_pattern, script.string, re.DOTALL)
            
            for match in matches:
                try:
                    json_obj = json.loads(match.group(1))
                    results.append(json_obj)
                except json.JSONDecodeError:
                    continue
    
    return results


def load_html_and_extract_json(file_path: str | Path, variable_name: Optional[str] = None) -> Optional[Dict[str, Any]]:
    """
    Load an HTML file and extract JSON variable from <script> tag.
    
    Args:
        file_path: Path to the HTML file
        variable_name: Optional name of the variable to extract
    
    Returns:
        The parsed JSON object, or None if not found
    
    Example:
        data = load_html_and_extract_json("page.html", "myData")
    """
    file_path = Path(file_path)
    
    if not file_path.exists():
        raise FileNotFoundError(f"HTML file not found: {file_path}")
    
    with open(file_path, 'r', encoding='utf-8') as f:
        html_content = f.read()
    
    return extract_json_from_script_tag(html_content, variable_name)


# Example usage
if __name__ == "__main__":
    # Example HTML content
    example_html = """
    <!DOCTYPE html>
    <html>
    <head>
        <title>Example</title>
    </head>
    <body>
        <script>
            var myData = {
                "name": "John",
                "age": 30,
                "items": [1, 2, 3]
            };
        </script>
    </body>
    </html>
    """
    
    # Extract the JSON
    data = extract_json_from_script_tag(example_html, "myData")
    print("Extracted data:", data)
    
    # Or extract any JSON
    data = extract_json_from_script_tag(example_html)
    print("Any JSON found:", data)

File: frontend/lib/htmlParser.ts
/**
 * Utility functions for parsing HTML files and extracting JSON from <script> tags.
 */

/**
 * Extract JSON variable from a <script> tag in HTML content.
 * 
 * @param htmlContent - The HTML content as a string
 * @param variableName - Optional name of the variable to extract (e.g., "window.data" or "data")
 *                      If undefined, will try to extract any JSON object from script tags
 * @returns The parsed JSON object, or null if not found
 * 
 * @example
 * // Extract a specific variable: var data = {...};
 * const data = extractJsonFromScriptTag(html, "data");
 * 
 * // Extract window.myData = {...};
 * const data = extractJsonFromScriptTag(html, "window.myData");
 * 
 * // Extract any JSON object from script tags
 * const data = extractJsonFromScriptTag(html);
 */
export function extractJsonFromScriptTag(
  htmlContent: string,
  variableName?: string
): any | null {
  if (variableName) {
    // Method 1: Extract specific variable using regex
    // Pattern matches: var variable_name = {...}; or window.variable_name = {...};
    const patterns = [
      // var variable_name = {...};
      new RegExp(`\\bvar\\s+${escapeRegex(variableName)}\\s*=\\s*(\\{.*?\\})\\s*;`, 's'),
      // let variable_name = {...};
      new RegExp(`\\blet\\s+${escapeRegex(variableName)}\\s*=\\s*(\\{.*?\\})\\s*;`, 's'),
      // const variable_name = {...};
      new RegExp(`\\bconst\\s+${escapeRegex(variableName)}\\s*=\\s*(\\{.*?\\})\\s*;`, 's'),
      // window.variable_name = {...};
      new RegExp(`window\\.${escapeRegex(variableName)}\\s*=\\s*(\\{.*?\\})\\s*;`, 's'),
      // variable_name = {...};
      new RegExp(`\\b${escapeRegex(variableName)}\\s*=\\s*(\\{.*?\\})\\s*;`, 's'),
    ];

    for (const pattern of patterns) {
      const match = htmlContent.match(pattern);
      if (match && match[1]) {
        try {
          return JSON.parse(match[1]);
        } catch (e) {
          continue;
        }
      }
    }

    // Method 2: Using DOMParser to find script tags and search within them
    const parser = new DOMParser();
    const doc = parser.parseFromString(htmlContent, 'text/html');
    const scripts = doc.querySelectorAll('script');

    for (const script of scripts) {
      if (script.textContent) {
        for (const pattern of patterns) {
          const match = script.textContent.match(pattern);
          if (match && match[1]) {
            try {
              return JSON.parse(match[1]);
            } catch (e) {
              continue;
            }
          }
        }
      }
    }
  } else {
    // Extract any JSON object from script tags
    const parser = new DOMParser();
    const doc = parser.parseFromString(htmlContent, 'text/html');
    const scripts = doc.querySelectorAll('script');

    for (const script of scripts) {
      if (script.textContent) {
        // Try to find JSON objects in the script content
        // Look for patterns like {...} or [{...}]
        const jsonPatterns = [
          /(\{[^{}]*(?:\{[^{}]*\}[^{}]*)*\})/s, // Nested objects
          /(\[[^\]]*\])/s, // Arrays
        ];

        for (const pattern of jsonPatterns) {
          const matches = script.textContent.matchAll(pattern);
          for (const match of matches) {
            if (match[1]) {
              try {
                return JSON.parse(match[1]);
              } catch (e) {
                continue;
              }
            }
          }
        }
      }
    }
  }

  return null;
}

/**
 * Extract all JSON objects from <script> tags in HTML content.
 * 
 * @param htmlContent - The HTML content as a string
 * @returns Array of all parsed JSON objects found in script tags
 */
export function extractAllJsonFromScriptTags(htmlContent: string): any[] {
  const results: any[] = [];
  const parser = new DOMParser();
  const doc = parser.parseFromString(htmlContent, 'text/html');
  const scripts = doc.querySelectorAll('script');

  for (const script of scripts) {
    if (script.textContent) {
      // Find all JSON-like objects
      const jsonPattern = /(\{[^{}]*(?:\{[^{}]*\}[^{}]*)*\})/s;
      const matches = script.textContent.matchAll(jsonPattern);

      for (const match of matches) {
        if (match[1]) {
          try {
            const jsonObj = JSON.parse(match[1]);
            results.push(jsonObj);
          } catch (e) {
            continue;
          }
        }
      }
    }
  }

  return results;
}

/**
 * Load an HTML file and extract JSON variable from <script> tag.
 * 
 * @param file - File object or file path
 * @param variableName - Optional name of the variable to extract
 * @returns Promise that resolves to the parsed JSON object, or null if not found
 * 
 * @example
 * const file = event.target.files[0];
 * const data = await loadHtmlAndExtractJson(file, "myData");
 */
export async function loadHtmlAndExtractJson(
  file: File | string,
  variableName?: string
): Promise<any | null> {
  let htmlContent: string;

  if (typeof file === 'string') {
    // If it's a URL or path, fetch it
    const response = await fetch(file);
    htmlContent = await response.text();
  } else {
    // If it's a File object, read it
    htmlContent = await file.text();
  }

  return extractJsonFromScriptTag(htmlContent, variableName);
}

/**
 * Helper function to escape special regex characters
 */
function escapeRegex(str: string): string {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

// Example usage (for Node.js/server-side, you might need to use a different HTML parser)
// For browser/client-side, the above functions work with DOMParser

